File: SurgSignDesktop/SurgSignDesktop/Models/TransferPayload.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace SurgSignDesktop.Models;

/// <summary>
/// QR transfer format v2 — matches the Swift <c>TransferPayload</c> exactly.
///
/// <para>
/// Keys are single-letter and alphabetically ordered via <see cref="JsonPropertyOrderAttribute"/>
/// to reproduce the output of Swift's <c>JSONEncoder</c> with <c>.sortedKeys</c>.
/// </para>
/// </summary>
public sealed class TransferPayload
{
    /// <summary>Shift label — <c>"AM"</c> or <c>"PM"</c>.</summary>
    [JsonPropertyName("from")]
    [JsonPropertyOrder(0)]
    public string From { get; set; } = "AM";

    /// <summary>Patient DTOs.</summary>
    [JsonPropertyName("pts")]
    [JsonPropertyOrder(1)]
    public List<PatientDTO> Pts { get; set; } = new();

    /// <summary>Service name.</summary>
    [JsonPropertyName("svc")]
    [JsonPropertyOrder(2)]
    public string Svc { get; set; } = string.Empty;

    /// <summary>Schema version — always <c>2</c>.</summary>
    [JsonPropertyName("v")]
    [JsonPropertyOrder(3)]
    public int V { get; set; } = 2;
}

/// <summary>
/// Patient DTO — single-letter keys (alphabetically ordered).
/// </summary>
public sealed class PatientDTO
{
    [JsonPropertyName("al")]
    [JsonPropertyOrder(0)]
    public string Al { get; set; } = string.Empty;

    [JsonPropertyName("cs")]
    [JsonPropertyOrder(1)]
    public string Cs { get; set; } = string.Empty;

    [JsonPropertyName("dx")]
    [JsonPropertyOrder(2)]
    public string Dx { get; set; } = string.Empty;

    [JsonPropertyName("id")]
    [JsonPropertyOrder(3)]
    public string Id { get; set; } = string.Empty;

    /// <summary>Tasks array — Swift uses <c>"is"</c> as the CodingKey.</summary>
    [JsonPropertyName("is")]
    [JsonPropertyOrder(4)]
    public List<TaskDTO> Is { get; set; } = new();

    [JsonPropertyName("m")]
    [JsonPropertyOrder(5)]
    public string M { get; set; } = string.Empty;

    [JsonPropertyName("n")]
    [JsonPropertyOrder(6)]
    public string N { get; set; } = string.Empty;

    [JsonPropertyName("nt")]
    [JsonPropertyOrder(7)]
    public string Nt { get; set; } = string.Empty;

    [JsonPropertyName("pr")]
    [JsonPropertyOrder(8)]
    public string Pr { get; set; } = string.Empty;

    [JsonPropertyName("px")]
    [JsonPropertyOrder(9)]
    public string Px { get; set; } = string.Empty;

    [JsonPropertyName("r")]
    [JsonPropertyOrder(10)]
    public string R { get; set; } = string.Empty;

    [JsonPropertyName("s")]
    [JsonPropertyOrder(11)]
    public string S { get; set; } = string.Empty;

    [JsonPropertyName("sd")]
    [JsonPropertyOrder(12)]
    public string Sd { get; set; } = string.Empty;
}

/// <summary>
/// Task DTO — single-letter keys, alphabetically ordered (<c>d</c>, <c>id</c>, <c>t</c>).
/// </summary>
public sealed class TaskDTO
{
    /// <summary>Done flag — 1 if done, 0 if not.</summary>
    [JsonPropertyName("d")]
    [JsonPropertyOrder(0)]
    public int D { get; set; }

    [JsonPropertyName("id")]
    [JsonPropertyOrder(1)]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("t")]
    [JsonPropertyOrder(2)]
    public string T { get; set; } = string.Empty;
}

File: SurgSign-iOS/SurgSign/Models/TransferPayload.swift
import Foundation

// MARK: - Transfer Payload

/// A pure Codable struct representing the v2 QR transfer format.
/// Matches the PWA's JSON encoding exactly for cross-platform compatibility.
/// This struct has no SwiftData dependency — ViewModels bridge between
/// SwiftData `@Model` objects and this DTO layer.
struct TransferPayload: Codable, Sendable {
    let v: Int
    let svc: String
    let from: String
    let pts: [PatientDTO]

    // MARK: - Patient DTO

    struct PatientDTO: Codable, Sendable {
        let id: String
        let n: String    // "Last, First"
        let m: String    // MRN
        let r: String    // room
        let s: String    // surgeon
        let dx: String   // diagnosis
        let px: String   // procedure
        let sd: String   // surgery date (ISO-8601 date string)
        let cs: String   // code status raw value
        let al: String   // allergies
        let pr: String   // priority raw value
        let nt: String   // notes
        let issues: [TaskDTO]

        enum CodingKeys: String, CodingKey {
            case id, n, m, r, s, dx, px, sd, cs, al, pr, nt
            case issues = "is"
        }

        // MARK: - Task DTO

        struct TaskDTO: Codable, Sendable {
            let id: String
            let t: String   // text
            let d: Int      // 0 = not done, 1 = done
        }
    }

    // MARK: - Factory

    /// Build a transfer payload from an array of `PatientTransferData`.
    static func build(
        patients: [PatientTransferData],
        serviceName: String,
        shift: String
    ) -> TransferPayload {
        let dtos = patients.map { p in
            PatientDTO(
                id: p.id,
                n: p.name,
                m: p.mrn,
                r: p.room,
                s: p.surgeon,
                dx: p.diagnosis,
                px: p.procedure,
                sd: p.surgeryDate,
                cs: p.codeStatus,
                al: p.allergies,
                pr: p.priority,
                nt: p.notes,
                issues: p.tasks.map { task in
                    PatientDTO.TaskDTO(
                        id: task.id,
                        t: task.text,
                        d: task.isDone ? 1 : 0
                    )
                }
            )
        }
        return TransferPayload(v: 2, svc: serviceName, from: shift, pts: dtos)
    }

    // MARK: - JSON Helpers

    /// Encode this payload to compact JSON data (no extra whitespace).
    func jsonData() throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return try encoder.encode(self)
    }

    /// Encode this payload to a compact JSON string.
    func jsonString() throws -> String {
        let data = try jsonData()
        guard let string = String(data: data, encoding: .utf8) else {
            throw TransferPayloadError.encodingFailed
        }
        return string
    }

    /// Decode a `TransferPayload` from JSON data.
    static func from(jsonData data: Data) throws -> TransferPayload {
        let decoder = JSONDecoder()
        return try decoder.decode(TransferPayload.self, from: data)
    }

    /// Decode a `TransferPayload` from a JSON string.
    static func from(jsonString string: String) throws -> TransferPayload {
        guard let data = string.data(using: .utf8) else {
            throw TransferPayloadError.decodingFailed
        }
        return try from(jsonData: data)
    }
}

// MARK: - PatientTransferData

/// A lightweight, SwiftData-free struct used to pass patient information
/// into the services layer. ViewModels create these from `@Model` objects.
struct PatientTransferData: Sendable {
    let id: String
    let name: String
    let mrn: String
    let room: String
    let surgeon: String
    let diagnosis: String
    let procedure: String
    let surgeryDate: String
    let codeStatus: String
    let allergies: String
    let priority: String
    let notes: String
    let tasks: [TaskTransferData]

    struct TaskTransferData: Sendable {
        let id: String
        let text: String
        let isDone: Bool
    }
}

// MARK: - Errors

enum TransferPayloadError: LocalizedError {
    case encodingFailed
    case decodingFailed

    var errorDescription: String? {
        switch self {
        case .encodingFailed:
            return "Failed to encode the transfer payload to JSON."
        case .decodingFailed:
            return "Failed to decode the transfer payload from JSON."
        }
    }
}